### PR TITLE
[OAP-Cache][POAE7-348]Fix OAP not apply ReuseExchange Optimization

### PR DESCRIPTION
**Pull request by jikunshang**
## What changes were proposed in this pull request?

OAP  will not apply ReuseExchange optimization due to `OAPFileFormat` didn't override `equals`, `hashcode`, `shortName`, `toString` method. 

Related issue: [1110](https://github.com/Intel-bigdata/OAP/issues/1110), [1423](https://github.com/Intel-bigdata/OAP/issues/1423)

## How was this patch tested?

add a UT, and run TPC-DS query64 , get about more than 2x performance gain(632s -> 289s, lower better).

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

**Comment by LuciferYang** [comment]
`equals` and `hashcode` is the key point

**Comment by LuciferYang** [comment]
```
- test OAP accumulators on Parquet when join *** FAILED ***
  200 did not equal 100 (OapMetricsSuite.scala:72)
```
1 Test failed, @jikunshang

File: oap-cache/oap/src/test/scala/org/apache/spark/sql/execution/datasources/oap/OapMetricsSuite.scala [hunk 1]
```diff
@@ -163,11 +162,12 @@ class OapMetricsSuite extends QueryTest with SharedOapContext with BeforeAndAfte
     checkAnswer(df, Row(1, "this is row 1") :: Nil)
 
     // rowsRead = 2 && rowsSkipped = 2 * 99
+    // ReuseExchange optimization
     val fileFormats = getOapFileFormat(df.queryExecution.sparkPlan)
     fileFormats.foreach(f => assert(f.orNull != null))
     assertAccumulators(fileFormats,
-      Some(RowMetrics(2 * 100, 0, 2 * 1, 2 * 99, 0, 0)),
-      Some(TaskMetrics(2 * 3, 0, 2 * 3, 0, 0)))
+      Some(RowMetrics(100, 0, 1, 99, 0, 0)),
```
Comment (LuciferYang) [inline oap-cache/oap/src/test/scala/org/apache/spark/sql/execution/datasources/oap/OapMetricsSuite.scala:170]: I don't think this is the correct fix, there is no shuffle reuse scenario triggered

**Comment by LuciferYang** [comment]
@jikunshang On the other hand, `OapFileFormat` and subclass are all stateful such as they have `oapmeta`, OapFileFormat holds different meta instance where read different tables, so just judge  the equality of `OapFileFormat` instance  has potential risks as following:

1. A OapFileFormat instance init table A meta will equals another OapFileFormat instance init table B meta

1. One OapFileFormat instance will equals one OptimizedOrcFileFormat instance

**Comment by jikunshang** [comment]
> @jikunshang On the other hand, `OapFileFormat` and subclass are all stateful such as they have `oapmeta`, OapFileFormat holds different meta instance where read different tables, so just judge the type equality of `OapFileFormat` instance has potential risks as following:
> 
> 1. A OapFileFormat instance init table A meta will equals another OapFileFormat instance init table B meta
> 2. A OapFileFormat instance will equals a OptimizedOrcFileFormat instance

@LuciferYang Thanks for your comments! 

I noticed that `ParquetFileFormat` also implemnet `equal` method in a naive [way](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala#L72). Will Parquet have the 1st risk too? 

About risk 2, I override `equals` in `OptimizedOrcFileFormat` and `OptimizedParquetFileFormat` , I think this will not be a risk, correct me if I'm wrong.

**Comment by LuciferYang** [comment]
> I noticed that ParquetFileFormat also implemnet equal method in a naive way. Will Parquet have the 1st risk too?
No,`ParquestFileFormat` is Stateless

**Comment by LuciferYang** [comment]
> About risk 2, I override equals in OptimizedOrcFileFormat and OptimizedParquetFileFormat , I think this will not be a risk, correct me if I'm wrong.
@jikunshang You can test manually as following:
```
println(new OptimizedOrcFileFormat().equals(new OapFileFormat))
println(new OapFileFormat().equals(new OptimizedOrcFileFormat))
```
The results are false , true. `a == b` but `b != a` ? Isn't it strange? ,

**Comment by jikunshang** [comment]
> > I noticed that ParquetFileFormat also implemnet equal method in a naive way. Will Parquet have the 1st risk too?
> 
> No,`ParquetFileFormat` is Stateless

So we should judge `OapDataFileMeta` in `equals` method?  I heard from @zhixingheyi-tian that you are trying to refactor this part. Do you have any progress or code can share with us?

**Comment by LuciferYang** [comment]
@jikunshang #1337 , there is a 2.4 base branch, no 3.0 base branch

**Comment by jikunshang** [comment]
@LuciferYang I have removed fix for `OAPFileFormat`, do you think it make sense now? Any other suggestion?

**Comment by LuciferYang** [comment]
+1

**Comment by LuciferYang** [comment]
@winningsix @jikunshang master does not need this patch?

**Comment by jikunshang** [comment]
@LuciferYang I have cherry-picked and pushed this patch to master branch.

**Pull request merged**